fathom/src/core/binary.rs

```diff
@@ -39,42 +43,26 @@ impl<'arena, 'env> Context<'arena, 'env> {
         &mut self,
         reader: &mut dyn SeekRead,
         format: ArcValue<'arena>,
-    ) -> io::Result<HashMap<u64, RefData<'arena>>> {
+    ) -> io::Result<HashMap<u64, Vec<ParsedRef<'arena>>>> {
         let initial_pos = reader.stream_position()?;
-        let mut refs = HashMap::<_, RefData<'_>>::new();
+        let mut refs = HashMap::<_, Vec<ParsedRef<'_>>>::new();
 
         // Parse the entrypoint from the beginning start of the binary data
         self.pending_formats.push((0, format));
 
-        // NOTE: This could lead to non-termination if we aren't careful!
         while let Some((pos, format)) = self.pending_formats.pop() {
-            use std::collections::hash_map::Entry;
-
-            let format_repr = self.elim_context().apply_repr(&format);
-
-            match refs.entry(pos) {
-                Entry::Occupied(entry) => {
-                    let RefData { r#type, .. } = entry.get();
-                    // Ensure that the format's representation type matches the
-                    // type of the stored reference.
-                    if !self.conversion_context().is_equal(r#type, &format_repr) {
-                        return Err(io::Error::new(
-                            io::ErrorKind::Other,
-                            "ref is occupied by an incompatible type",
-                        ));
-                    }
-                }
-                Entry::Vacant(entry) => {
-                    // Seek to current current ref location
-                    reader.seek(SeekFrom::Start(pos))?;
-                    // Parse the data at that location
-                    let expr = self.read_format(reader, &format)?;
-                    // Record the data in the `refs` hashmap
-                    entry.insert(RefData {
-                        r#type: format_repr,
-                        expr,
-                    });
-                }
+            let parsed_refs = refs.entry(pos).or_insert(Vec::with_capacity(1));
+
+            if (parsed_refs.iter())
+                .find(|r| self.conversion_context().is_equal(&r.format, &format))
```

Comment:
It would be nice to make this more efficient. This would be easier if we could somehow assign a unique id to formats at compile time. This is complicated due to:

- runtime data dependencies in format descriptions 
- the fact that formats can be constructed at runtime (I think #316 might resolve this)

Reply:
Now I'm curious if Kaitai allows instances to be parametrised values 🤔

Reply:
Seems like no, Kaitai does not allow instances to have parameters, as far as I can see. Nor can they handle mutual recursion of the font tables (they seem to skip the tables where this comes up, like `htmx`?).